pandas/core/groupby/groupby.py

```diff
@@ -768,7 +770,17 @@ def _try_cast(self, result, obj, numeric_only=False):
                 # The function can return something of any type, so check
                 # if the type is compatible with the calling EA.
                 try:
+                    tz_dtype = None
+                    if is_datetime64tz_dtype(dtype):
+                        # GH 23683
+                        # Prior results were generated in UTC. Ensure
+                        # We localize to UTC first before converting to
+                        # the target timezone
+                        tz_dtype = dtype
+                        dtype = DatetimeTZDtype(tz='UTC')
                     result = obj._values._from_sequence(result, dtype=dtype)
+                    if tz_dtype is not None:
```

Comment:
Looks generally good. Is this something you see us needing to do elsewhere besides just this block in groupby (no idea myself, just asking)

Reply:
It appears that `_try_cast` is the most downstream recipient of results from the cython dispatch. Not too familiar with the groupby ops though.

But this generally needs to be down when datetimetz data goes through cython routines.